cmd/consul-dataplane/merge.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	"strings"
+
+	"github.com/hashicorp/consul-dataplane/pkg/consuldp"
+)
+
+// mergeConfigs takes in a couple of configs(c1&c2) and applies
+// c2 on to c1 based on the following conditions
+//
+// 1. If an attribute of c2 is a string/integer/bool, then it is applied to the
+// same field in c1 if it holds non default value for the field
+// 2. If an attribute of c2 is a slice/map, then it overrides the same field in c1
+func mergeConfigs(c1, c2 *consuldp.Config) {
+	mergeConsulConfigs(c1.Consul, c2.Consul)
+	mergeServiceConfigs(c1.Service, c2.Service)
+	mergeTelemetryConfigs(c1.Telemetry, c2.Telemetry)
+	mergeLoggingConfigs(c1.Logging, c2.Logging)
+	mergeEnvoyConfigs(c1.Envoy, c2.Envoy)
+	mergeXDSServerConfigs(c1.XDSServer, c2.XDSServer)
+	mergeDNSServerConfigs(c1.DNSServer, c2.DNSServer)
+}
+
+func mergeConsulConfigs(c1, c2 *consuldp.ConsulConfig) {
+	if c2 == nil {
+		return
+	}
+
+	if c2.Addresses != "" {
+		c1.Addresses = c2.Addresses
+	}
+
+	if c2.GRPCPort != int(0) && c2.GRPCPort != DefaultGRPCPort {
+		c1.GRPCPort = c2.GRPCPort
+	}
+
+	if c2.ServerWatchDisabled {
+		c1.ServerWatchDisabled = true
+	}
+
+	if c2.Credentials != nil {
+		if c2.Credentials.Type != "" {
+			c1.Credentials.Type = c2.Credentials.Type
+		}
+
+		if c2.Credentials.Login.AuthMethod != "" {
+			c1.Credentials.Login.AuthMethod = c2.Credentials.Login.AuthMethod
+		}
+
+		if c2.Credentials.Login.BearerToken != "" {
+			c1.Credentials.Login.BearerToken = c2.Credentials.Login.BearerToken
+		}
+
+		if c2.Credentials.Login.BearerTokenPath != "" {
+			c1.Credentials.Login.BearerTokenPath = c2.Credentials.Login.BearerTokenPath
+		}
+
+		if c2.Credentials.Login.Datacenter != "" {
+			c1.Credentials.Login.Datacenter = c2.Credentials.Login.Datacenter
+		}
+
+		if c2.Credentials.Login.Namespace != "" {
+			c1.Credentials.Login.Namespace = c2.Credentials.Login.Namespace
+		}
+
+		if c2.Credentials.Login.Partition != "" {
+			c1.Credentials.Login.Partition = c2.Credentials.Login.Partition
+		}
+
+		if c2.Credentials.Login.Meta != nil {
+			c1.Credentials.Login.Meta = c2.Credentials.Login.Meta
+		}
+
+		if c2.Credentials.Static.Token != "" {
+			c1.Credentials.Static.Token = c2.Credentials.Static.Token
+		}
+	}
+
+	if c2.TLS != nil {
+		if c2.TLS.Disabled {
+			c1.TLS.Disabled = true
```

Comment:
I think - if `TLS.Disabled` is true in config file but false in flag, ideally final config should be false right?.
This code will not set it to false.